File: update_db.js
import path from 'path';
import { createReadStream } from 'fs';
import { EOL } from 'os';
import logger from './logger';
import db from './db';

const csvCheckedFilePath = path.resolve(__dirname, 'check_success.csv');

const migrateDB = async (data, transaction) => {
    // using literal, write a query to update all the data in one go using case statement
    const query = `
    UPDATE files
    SET url = 
        CASE
            ${data.map(([public_id, cloudinary_url, bunny_url]) => `WHEN "cloudinaryId" = '${public_id}' THEN '${bunny_url}'`)}
        END
    WHERE "cloudinaryId" IN (${data.map(([public_id, cloudinary_url, bunny_url]) => `'${public_id}'`)});
    `;

    console.log(query);
    // await db.query(query, {
    //     type: db.QueryTypes.UPDATE,
    //     transaction,
    // });
}

const parseChunk = async (text, index, transaction) => {
    if (index === 0) {
        // First chunk contains the header
        let headerLine = text.substring(0, text.indexOf(EOL));

        text = text.replace(headerLine+EOL, '');
    }

    const res = [];

    let lines = text.split(EOL);
    for(let line of lines) {
        let values = line.split(',');
        const [public_id, cloudinary_url, bunny_url] = values;

        if (!public_id || !cloudinary_url || !bunny_url) {
            logger.error('Invalid CSV line:', line);
            continue;
        }

        res.push([public_id, cloudinary_url, bunny_url]);
    }

    await migrateDB(res, transaction);
}

export const csvReader = async (transaction) => {
    return new Promise((resolve, reject) => {
        let chunkCount = 0;
        const readStream = createReadStream(csvCheckedFilePath);

        readStream.on('data', async chunk => {
            const text = chunk.toString();
            await parseChunk(text, chunkCount, transaction);
            chunkCount++;
        });

        readStream.on('end', () => {
            resolve();
        });

        readStream.on('error', error => {
            reject(error);
        });
    })
}

const readCsvInBatches = async () => {
    const transaction = db.transaction();
    try {
        await csvReader(transaction);
        logger.info('CSV processing completed.');
        await transaction.commit();
    } catch (error) {
        logger.error('Error processing CSV:', error);
    }
}

readCsvInBatches();

File: update_db.py
import os
import sys
import pandas as pd
import pathlib
import sys
from dotenv import load_dotenv
from sqlalchemy import create_engine, text

load_dotenv()

# FILENAME = "check_success.csv"
# FULLPATH = os.path.join(pathlib.Path(__file__).parent.absolute(), FILENAME)

FULLPATH = None

DB_URL = os.getenv("PY_POSTGRES_URL")

engine = create_engine(DB_URL)


def create_query(data):
    case_statement = []
    where_statement = []

    for _, row in data.iterrows():
        case_statement.append(
            "WHEN \"cloudinaryId\" = '{}' THEN '{}'".format(
                row["public_id"].replace("'", "''"),
                row["bunny_url"].replace("'", "''"),
            )
        )

        where_statement = [
            "'{}'".format(row["public_id"].replace("'", "''"))
            for _, row in data.iterrows()
        ]

    query = """
    UPDATE files
    SET url = 
        CASE
            {}
        END
    WHERE "cloudinaryId" IN ({});
    """.format(
        " ".join(case_statement), ", ".join(where_statement)
    )

    return query


def read_file_in_chunks(file_path, chunk_size=500):
    """
    Generator function to read a file in chunks.

    :param file_path: Path to the file to be read.
    :param chunk_size: Number of bytes to read at a time (default is 1024).
    :yield: Chunk of the file.
    """
    for chunk in pd.read_csv(file_path, chunksize=chunk_size):
        yield chunk


def main():
    FULLPATH = sys.argv[1]

    if not os.path.exists(FULLPATH):
        print("File not found")
        sys.exit(1)

    total = 0

    # Create a transaction
    with engine.begin() as conn:
        # Read file in chunks of 500 lines
        for chunk in read_file_in_chunks(FULLPATH):
            query = create_query(chunk)
            query = text(query)
            total += 500
            print("Executing query", query)
            print("Total rows updated", total)
            # Execute the query
            conn.execute(query)

        # WARN: This is just for testing purposes
        # query = text("SELECT * FROM files limit 10;")
        # result = conn.execute(query)
        # for row in result:
        #     print(row)

        # conn.rollback()

    print("Done")


if __name__ == "__main__":
    main()
